rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+    ) -> Result<Self, SymCryptError> {
+        let n_bits_mod = modulus_buffer.len() as u32;
+        let rsa_key = allocate_rsa(0, n_bits_mod*8)?;
+        let u64_pub_exp = load_msb_first_u64(pub_exp)?;
+        unsafe {
+            // SAFETY: FFI calls
+            // When only setting the public key, ppPrimes, pcbPrimes and nPrimes can be NULL, NULL and 0
+            match symcrypt_sys::SymCryptRsakeySetValue(
+                modulus_buffer.as_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                [u64_pub_exp].as_ptr(), // This array has a length of 1.
+                1 as symcrypt_sys::UINT32,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0,
+                NumberFormat::MSB.to_num_format(),
+                key_usage.to_flag(), // !Review: can set this sign and encrypt
+                rsa_key,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaPublicKey {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `size_of_modulus()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the modulus of the key.
+    pub fn size_of_modulus(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeySizeofModulus(self.inner)
+        }
+    }
+
+    /// `size_of_public_exponent()` returns a `u32` representing the (tight) size in bytes of a byte array big enough to store
+    /// the public exponent of the key.
+    pub fn size_of_public_exponent(&self) -> u32 {
+        unsafe {
+            // SAFETY: FFI calls
+            // Only one public exponent is supported, so the only valid index is 0.
+            symcrypt_sys::SymCryptRsakeySizeofPublicExponent(self.inner, 0)
+        }
+    }
+    /// `key_usage` returns the intended usage for the RSA public key.
+    pub fn key_usage(&self) -> RsaKeyUsage {
+        self.key_usage
+    }
+
+    /// `key_usage` gives crate access to the 
+    pub(crate) fn inner(&self) -> symcrypt_sys::PSYMCRYPT_RSAKEY  {
+        self.inner
+    }
+}
+
+impl Drop for RsaPublicKey {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeyFree(self.inner);
+        }
+    }
+}
+
+/// A trait for RSA signing.
+pub trait RsaSign {
+    /// Signs a message using the RSA private key.
+    ///
+    /// # Parameters
+    ///
+    /// - `message`: A byte slice representing the message to be signed.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(Vec<u8>)`: A vector of bytes containing the signature.
+    /// - `Err(String)`: An error message if the signing operation fails.
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, String>;
+}
+
+/// A trait for RSA signature verification.
```

Comment:
I'm guessing these traits will be implemented with various padding schemes in the OAEP, PKCS1 and PSS files?